File: backend/app/schemas/disbursement.py
"""
Pydantic Schemas for Loan Disbursement Workflow.
"""

from datetime import datetime
from decimal import Decimal
import uuid
from pydantic import BaseModel, ConfigDict


class DisbursementInitiateRequest(BaseModel):
    remarks: str | None = None


class DisbursementConfirmRequest(BaseModel):
    remarks: str | None = None


class DisbursementResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    disbursement_id: uuid.UUID
    application_id: uuid.UUID
    application_number: str
    approved_amount: Decimal
    net_disbursement_amount: Decimal
    disbursement_reference: str
    destination_account_summary: str | None
    status: str
    application_status: str
    initiated_at: datetime
    completed_at: datetime | None


class DisbursementDetailResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    application_id: uuid.UUID
    application_number: str
    application_status: str
    approved_amount: Decimal
    net_disbursement_amount: Decimal
    selected_offer_id: uuid.UUID | None
    interest_rate: Decimal | None
    tenure_months: int | None
    emi: Decimal | None
    processing_fee: Decimal | None
    gst: Decimal | None
    total_interest: Decimal | None
    total_repayment: Decimal | None
    approval_date: datetime | None
    reviewed_by: str | None
    disbursement_id: uuid.UUID | None
    disbursement_reference: str | None
    disbursement_status: str | None
    destination_bank_name: str | None
    destination_account_last4: str | None
    destination_ifsc: str | None
    account_holder_name: str | None
    initiated_at: datetime | None
    completed_at: datetime | None
    failure_reason: str | None

File: frontend/src/types/disbursement.ts
/**
 * TypeScript Interfaces for Phase 7 Loan Disbursement Lifecycle.
 */

export type DisbursementStatus = 'PENDING' | 'INITIATED' | 'SUCCESS' | 'FAILED';

export interface DisbursementResponse {
  disbursement_id: string;
  application_id: string;
  application_number: string;
  approved_amount: number | string;
  net_disbursement_amount: number | string;
  disbursement_reference: string;
  destination_account_summary: string | null;
  status: DisbursementStatus;
  application_status: string;
  initiated_at: string;
  completed_at: string | null;
}

export interface DisbursementDetail {
  application_id: string;
  application_number: string;
  application_status: string;
  approved_amount: number;
  net_disbursement_amount: number;
  selected_offer_id: string | null;
  interest_rate: number | null;
  tenure_months: number | null;
  emi: number | null;
  processing_fee: number | null;
  gst: number | null;
  total_interest: number | null;
  total_repayment: number | null;
  approval_date: string | null;
  reviewed_by: string | null;
  disbursement_id: string | null;
  disbursement_reference: string | null;
  disbursement_status: DisbursementStatus | string | null;
  destination_bank_name: string | null;
  destination_account_last4: string | null;
  destination_ifsc: string | null;
  account_holder_name: string | null;
  initiated_at: string | null;
  completed_at: string | null;
  failure_reason: string | null;
}
